File: 2022/day02.py
from data.day02 import SAMPLE_TXT, INPUT_TXT
from enum import IntEnum


class Rps(IntEnum):
    ROCK = 1
    PAPER = 2
    SCISSORS = 3


class Points(IntEnum):
    LOSE = 0
    DRAW = 3
    WIN = 6


OutcomeDict = dict[str, tuple[int, int]]

# (A, B, C) = RPS, (X, Y, Z) = RPS
OUTCOMES1: OutcomeDict = {'A X': (Rps.ROCK, Points.DRAW),
                          'B X': (Rps.ROCK, Points.LOSE),
                          'C X': (Rps.ROCK, Points.WIN),
                          'A Y': (Rps.PAPER, Points.WIN),
                          'B Y': (Rps.PAPER, Points.DRAW),
                          'C Y': (Rps.PAPER, Points.LOSE),
                          'A Z': (Rps.SCISSORS, Points.LOSE),
                          'B Z': (Rps.SCISSORS, Points.WIN),
                          'C Z': (Rps.SCISSORS, Points.DRAW)}

# (A, B, C) = RPS, (X, Y, Z) = LDW
OUTCOMES2: OutcomeDict = {'A X': (Rps.SCISSORS, Points.LOSE),
                          'B X': (Rps.ROCK, Points.LOSE),
                          'C X': (Rps.PAPER, Points.LOSE),
                          'A Y': (Rps.ROCK, Points.DRAW),
                          'B Y': (Rps.PAPER, Points.DRAW),
                          'C Y': (Rps.SCISSORS, Points.DRAW),
                          'A Z': (Rps.PAPER, Points.WIN),
                          'B Z': (Rps.SCISSORS, Points.WIN),
                          'C Z': (Rps.ROCK, Points.WIN)}


def either_part(txt: str, outcomes: OutcomeDict):
    return sum(sum(outcomes[pair]) for pair in txt.splitlines())


def part1(txt: str):
    return either_part(txt, OUTCOMES1)


def part2(txt: str):
    return either_part(txt, OUTCOMES2)


##########################
if __name__ == '__main__':
    assert part1(SAMPLE_TXT) == 15
    assert part2(SAMPLE_TXT) == 12
    print(part1(INPUT_TXT))
    print(part2(INPUT_TXT))

File: 2022-kotlin/day02.kt
import day2_data.INPUT_TXT
import day2_data.SAMPLE_TXT

typealias OutcomeHash = HashMap<String, List<Int>>

val OUTCOMES1: OutcomeHash = hashMapOf(
    "A X" to listOf(1, 3),
    "B X" to listOf(1, 0),
    "C X" to listOf(1, 6),
    "A Y" to listOf(2, 6),
    "B Y" to listOf(2, 3),
    "C Y" to listOf(2, 0),
    "A Z" to listOf(3, 0),
    "B Z" to listOf(3, 6),
    "C Z" to listOf(3, 3)
)

val OUTCOMES2: OutcomeHash = hashMapOf(
    "A X" to listOf(3, 0),
    "B X" to listOf(1, 0),
    "C X" to listOf(2, 0),
    "A Y" to listOf(1, 3),
    "B Y" to listOf(2, 3),
    "C Y" to listOf(3, 3),
    "A Z" to listOf(2, 6),
    "B Z" to listOf(3, 6),
    "C Z" to listOf(1, 6)
)

private fun part1(txt: String): Int{
    return eitherPart(txt, OUTCOMES1)
}

private fun part2(txt: String): Int {
    return eitherPart(txt, OUTCOMES2)
}

private fun eitherPart(txt: String, outcomeHash: OutcomeHash): Int{
    return txt.lines().sumOf { outcomeHash[it]!!.sum() }
}

fun main() {
    assert(part1(SAMPLE_TXT) == 15)
    println(part1(INPUT_TXT))
    assert(part2(SAMPLE_TXT) == 12)
    println(part2(INPUT_TXT))
}